### PR TITLE
Completed Express notes api | Shawn Kost

### PR DESCRIPTION
**Pull request by shawnkost**
## Get Requests
image
image
image
image

## Post Requests
image
image
image

## Delete Requests
image
image
image
image

## Put Requests
image
image
image
image
image

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  if (Math.sign(ID) !== 1) {
+    return res.status(400).json({
+      error: "ID given was not a positive integer"
+    });
+  }
+  for (const index in JSONData.notes) {
+    if (ID === Number(index)) {
+      return res.json(JSONData.notes[index]);
+    }
+  }
```
Comment (scbowler) [inline express-notes-api/index.js:28]: You don't need a loop here, you can just use an `if` statement:

```JAVASCRIPT
if(JSONData.notes[ID]) {
  // Note exists send it to client
} else {
  // Note does not exist send an error
}
```

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  res.status(404).json({
+    error: "ID does not exist"
+  });
+})
+
+app.post("/api/notes", (req, res) => {
+  const content = req.body;
+  const newID = JSONData.nextId++;
+  if (Object.entries(content).length === 0) {
```
Comment (scbowler) [inline express-notes-api/index.js:37]: Here you just need to check if `content` has a `content` property:

```JAVASCRIPT
if(content.content) {
  // Content exists make a new note
} else {
  // No content supplied by client, send an error
}
```

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+})
+
+app.delete("/api/notes/:id", (req, res) => {
+  const ID = Number(req.params.id);
+  if (Math.sign(ID) !== 1) {
+    return res.status(400).json({
+      error: "ID given was not a positive integer"
+    });
+  }
+  if (!JSONData.notes.hasOwnProperty(ID)) {
```
Comment (scbowler) [inline express-notes-api/index.js:64]: You don't need to use `hasOwnProperty` just directly check if it exists or not:

```JAVASCRIPT
if(!JSONData.notes[ID]) {
```

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    return res.status(404).json({
+      error: `cannot find note with id ${ID}`
+    });
+  }
+  for (const index in JSONData.notes) {
+    if (ID === Number(index)) {
+      delete JSONData.notes[ID];
+    }
+  }
```
Comment (scbowler) [inline express-notes-api/index.js:73]: You don't need a loop here, you have already verified rather the ID is valid or not, just delete it now.

```JAVASCRIPT
delete JSONData.notes[ID];
```

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+app.put("/api/notes/:id", (req, res) => {
+  const content = req.body;
+  const ID = Number(req.params.id);
+  if (Math.sign(ID) !== 1) {
+    return res.status(400).json({
+      error: "ID given was not a positive integer"
+    });
+  }
+  if (Object.entries(content).length === 0) {
```
Comment (scbowler) [inline express-notes-api/index.js:94]: Here you should be specifically checking just for a `content` property, same as in the POST method.

```JAVASCRIPT
if(content.content) {
  // Content exists make a new note
} else {
  // No content supplied by client, send an error
}
```

File: express-notes-api/index.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    return res.status(400).json({
+      error: "content is a required field"
+    });
+  }
+  if (!JSONData.notes.hasOwnProperty(ID)) {
```
Comment (scbowler) [inline express-notes-api/index.js:99]: Same as above, just directly use the `ID` to check if it exists or not

```JAVASCRIPT
if(!JSONData.notes[ID]) {
```

**Review by scbowler — CHANGES_REQUESTED** [review]
Looking good, but some updates needed. This should help simplify your code a bit.

File: express-notes-api/index.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  const newID = JSONData.nextId++;
+  if (content.content) {
+    JSONData.notes[newID] = content;
+    JSONData.notes[newID].id = newID;
+  } else {
+    return res.status(400).json({
+      error: "content is a required field"
+    });
+  }
+  let newEntry = JSONData;
```
Comment (Zhiwutian) [inline express-notes-api/index.js:44]: combine this line and the one below it

**Review by Zhiwutian — APPROVED** [review]
Approved but update before merge!

**Pull request merged**